### PR TITLE
Compute the eigenvalues of the CANDECOMP/PARAFAC decomposition

### PR DESCRIPTION
**Pull request by cswiercz**
Some versions of the CP decomposition algorithm return the eigenvalues along
with appropriately normalized factor matrices. This change adds a flag
`with_eigenvalues` which will optionally return these normalized factors along
with an array of `rank` eigenvalues.

This commit also makes necessary modifications to the backend `norm` functions
to support computing norms across particular axes.

### CODE REVIEW

**Comment by cswiercz** [comment]
Working on the `norm` test coverage.

**Comment by cswiercz** [comment]
@JeanKossaifi Ready for review. Please let me know if you have any questions. I can add more CPDecomp tests if you like. For example, I can write tests where I compare output eigenvalues and factors to input eigenvalues and (normalized) factors. For small problems we should hit convergence in enough restarts. (As you know, the necessity of restarts are a sad part of CPDecomp.)

**Comment by JeanKossaifi** [comment]
Thanks for the pull request @cswiercz!

Maybe we could split it in two, and already merge the tests you added for the backend, the changes you made to the norm function and the prod function?

For the tests you mention, what would you change between the restarts?

Regarding the normalization of the factors, we could do it post-decomposition in a separate function  (this would avoid having parafac return a variable number of values)? I think @ahwillia also worked on something similar?

**Comment by ahwillia** [comment]
All my code is very simple and basic. But feel free to steal anything you like -- https://github.com/ahwillia/tensortools/blob/master/tensortools/kruskal.py#L8

File: tensorly/backend/mxnet_backend.py [hunk 1]
```diff
@@ -59,14 +59,17 @@ def ndim(tensor):
     return tensor.ndim
 
 def arange(start, stop=None, step=1.0):
-    return np.arange(start, stop, step)
+    return np.arange(start, stop, step, dtype="float64")
```
[the remark this answers is not in the document]
Comment (cswiercz) [reply tensorly/backend/mxnet_backend.py]: Of course! No problem.

File: tensorly/decomposition/candecomp_parafac.py [hunk 2]
```diff
@@ -6,48 +6,36 @@
...
 # Author: Jean Kossaifi <jean.kossaifi+tensors@gmail.com>
+# Author: Chris Swierczewski <cswiercz@gmail.com>
 
 # License: BSD 3 clause
 
 
-def parafac(tensor, rank, n_iter_max=100, init='svd', tol=10e-7,
-            random_state=None, verbose=False):
-    """CANDECOMP/PARAFAC decomposition via alternating least squares (ALS)
+def initialize_factors(tensor, rank, init='svd', random_state=None):
```
[the remark this answers is not in the document]
Comment (cswiercz) [reply tensorly/decomposition/candecomp_parafac.py]: I would argue that having a separate function *improves* the readability of `parafac()`. The contents more comfortably fit in one page. Factor initialization is, from what I understand, an area of interest (read: "art") in and of itself. It's clear where the initialization takes place and how to add different initialization options. Pus, it's self-documenting: in the `parafac()` code it's clear that the factors are being initialized in the first step.

All that argument being said, I can change it back for you if you insist. :)

File: tensorly/decomposition/candecomp_parafac.py [hunk 3]
```diff
@@ -60,22 +48,77 @@ def parafac(tensor, rank, n_iter_max=100, init='svd', tol=10e-7,
...
         for mode in range(tensor.ndim):
             pseudo_inverse = T.tensor(np.ones((rank, rank)))
             for i, factor in enumerate(factors):
                 if i != mode:
-                    pseudo_inverse[:] = pseudo_inverse*T.dot(factor.T, factor)
+                    pseudo_inverse *= T.dot(factor.T, factor)
             factor = T.dot(unfold(tensor, mode), khatri_rao(factors, skip_matrix=mode))
             factor = T.solve(pseudo_inverse.T, factor.T).T
+            if with_eigenvalues:
+                eigenvalues = T.norm(factor, axis=0)
```
[the remark this answers is not in the document]
Comment (cswiercz) [reply tensorly/decomposition/candecomp_parafac.py]: This is something I'm still trying to understand. I've looked at a variety of implementations of CPDecomp/PARAFAC and they all ignore the eigenvalues of all but the final factor. From what I've learned so far the normalizations of each factor, in a sense, propagates the eigenvalue information forward to the next computed factor. The final factor, in some sort of sense, contains the eigenvalue data by the end of the given iteration.

We need to compute the column norms, anyway, and scale the factor so we may as well overwrite. I can add a comment to make this more clear.

File: tensorly/decomposition/candecomp_parafac.py [hunk 4]
```diff
@@ -88,6 +131,8 @@ def parafac(tensor, rank, n_iter_max=100, init='svd', tol=10e-7,
                     print('converged in {} iterations.'.format(iteration))
                 break
 
+    if with_eigenvalues:
+        return factors, eigenvalues
```
[the remark this answers is not in the document]
Comment (cswiercz) [reply tensorly/decomposition/candecomp_parafac.py]: I did run some experiments where I computed the eigenvalues and normalized the factors after the fact and it didn't seem to be as stable as performing the normalization at each step. It seems like it should be possible based on the following,

```
a \otimes b \otimes c
=
(\lambda_a a_normal) \otimes (\lambda_b b_normal) \otimes (\lambda_c c_normal)
=
(\lambda_a \lambda_b \lambda_c) a_normal \otimes b_normal \otimes c_normal
```

but, again, the results were a bit unstable. Maybe I got something wrong...

**Comment by ahwillia** [comment]
Sorry to chime in - is eigenvalues really the right term here? E.g. what if the tensor isn't cubic? CANDECOMP/PARAFAC is still well defined, but in that case the normalizing scalars are more analogous to the singular values of a matrix, no?

**Comment by cswiercz** [comment]
> Maybe we could split it in two, and already merge the tests you added for the backend, the changes you made to the norm function and the prod function?

Sure. I'll create a new PR for the new functionality.

> For the tests you mention, what would you change between the restarts?

The factor initialization would change with each restart. ALS has poor local minima. Plot reconstruction error as a function of iteration count for various initial factors. For an 8x8x8 rank 8 tensor it's not uncommon to find factors that only achieve 1e-2 reconstruction error after many iterations. The "good" initializations lead to 1e-8 and better reconstruction errors after a couple hundred iterations.

So the solution, as I see it, is to either (a) hunt for a good random number generator seed that will produce an initialization leading to a global optimum in relatively few iterations or (b) run several restarts and hope that at least one of them hits the global min. I forget the name but I recall finding a paper that had some probabilistic results on how many restarts you should expect to need.

> Regarding the normalization of the factors, we could do it post-decomposition in a separate function (this would avoid having parafac return a variable number of values)? I think @ahwillia also worked on something similar?

I unfortunately read this and @ahwillia 's comments after responding to the code changes above. I'll see if I can reliably recover the expected eigenvalues from my experiments using this code.

On that note, would you mind if I add to the `parafac()` test suite? :)

**Comment by cswiercz** [comment]
> Sorry to chime in - is eigenvalues really the right term here? E.g. what if the tensor isn't cubic? CANDECOMP/PARAFAC is still well defined, but in that case the normalizing scalars are more analogous to the singular values of a matrix, no?

Please chime in!

So the term "eigenvalue" (and "eigenvector" for that matter) are most analogous to the symmetric case,

```
T = \sum_{i=1}^R \lambda_i v_i^{\otimes P}
```

where `R` is the rank and `P` is the order. In the generic case, Kolda and Bader just refer to them as the "weights". I don't know if there is a canonical name for these parameters.

**Comment by cswiercz** [comment]
@ahwillia I tried using your [`normalize_factors()`](https://github.com/ahwillia/tensortools/blob/master/tensortools/kruskal.py#L8) function and it seems to do the job just fine. I agree with @JeanKossaifi that it might be cleaner to "post-process" the eigenvalues...

Given all of the other comments on this code perhaps I this PR should be closed ***without*** merge.

**Comment by JeanKossaifi** [comment]
@cswiercz we can have a pull request to include your improvements on the backend, the tests you added, and the separate function for the initialization of the factors (I agree with your argument, it's cleaner and leads the way to other methods for init)?

@ahwillia would you be ok creating a pull request to add your normalize_factors function? Adding an example visualising the CP decomp would be amazing :D

**Comment by ahwillia** [comment]
Sure, I can create a PR in a few days. Let me know if you have any requests
/ ideas for where to put `normalize_factors`.

File: tensorly/decomposition/candecomp_parafac.py [hunk 5]
```diff
@@ -60,22 +48,77 @@ def parafac(tensor, rank, n_iter_max=100, init='svd', tol=10e-7,
...
     norm_tensor = T.norm(tensor, 2)
+    eigenvalues = None
 
     for iteration in range(n_iter_max):
         for mode in range(tensor.ndim):
             pseudo_inverse = T.tensor(np.ones((rank, rank)))
             for i, factor in enumerate(factors):
                 if i != mode:
-                    pseudo_inverse[:] = pseudo_inverse*T.dot(factor.T, factor)
+                    pseudo_inverse *= T.dot(factor.T, factor)
```
Comment (cswiercz) [inline tensorly/decomposition/candecomp_parafac.py]: What was the reasoning behind the use of `[:]`, by the way?

**Comment by cswiercz** [comment]
Hmmm...looks like I'll have to rebase. I'll take the opportunity to squash my changes as well.

File: tensorly/kruskal_tensor.py [hunk 6]
```diff
@@ -43,11 +53,9 @@ def kruskal_to_tensor(factors, eigenvalues=None):
     """
     shape = [factor.shape[0] for factor in factors]
     first_factor = factors[0]
-    remaining_factors = factors[1:]
-    if eigenvalues is not None:
-        full_tensor = T.dot(first_factor*eigenvalues, khatri_rao(remaining_factors).T)
-    else:
-        full_tensor = T.dot(first_factor, khatri_rao(remaining_factors).T)
+    if weights is not None:
+        first_factor *= eigenvalues
```
Comment (cswiercz) [inline tensorly/kruskal_tensor.py]: Should be `weights` instead of `eigenvalues`.

**Comment by JeanKossaifi** [comment]
Feel free to just open another pull request if easier :)

#10 by @ahwillia adds the normalization but needs the changes make to norm in this PR

File: tensorly/decomposition/candecomp_parafac.py (hunk 5 above)
Comment (JeanKossaifi) [reply tensorly/decomposition/candecomp_parafac.py]: In MXNet it is used for inplace operations.

**Comment by cswiercz** [comment]
Closed due to inactivity.

**Pull request closed without merging**